### ISSUE TITLE
Selectable/Highlightable text?

### ISSUE THREAD

**Issue by PhilNorfleet** [eKoopmans/html2pdf#56]
Wondering if its possible to have the text output be highlightable? Or would this require an external OCR solution?

**Comment by eKoopmans** [comment]
Yeah, right now the PDF is generated first as an image, so any text information is lost, unfortunately. Rendering the text straight to PDF would be ideal but there aren't really any client-side options for that.

**Comment by kotAPI** [comment]
Hey @eKoopmans , any new ideas on how do do it? I'd like to contribute and help out if you've any solutions in mind.

**Comment by eKoopmans** [comment]
Hey @kotAPI! I appreciate the offer. There's only one solution that comes to mind, and it's out of my league, but you're definitely welcome to give it a shot!

The de facto standard for html->PDF generation is [wkhtmltopdf](https://wkhtmltopdf.org/). It's a command-line tool that takes an HTML string and generates a PDF (or image). It's great - it's built on the WebKit rendering engine, so you get exactly what a true web browser would see. So you could, for instance, set things up server-side to receive HTML, run it through wkhtmltopdf, and return the PDF. There are already websites out there that do exactly that. Unfortunately, html2pdf is designed to be entirely client-side, which means it needs to be 100% JavaScript.

The interesting part is that wkhtmltopdf has an [open-source C library](https://github.com/wkhtmltopdf/wkhtmltopdf). My C is pretty rusty, but I know there are tools out there (like [Emscripten](http://kripken.github.io/emscripten-site/)) to compile C into JS. For instance, see [this comment](https://stackoverflow.com/a/4212680/4080966) about Emscripten being used to compile DOOM into JS.

So, it might be possible to create a JS version of wkhtmltopdf, which would be very useful, but probably a lot of work. Even if it was possible, the file size might end up being prohibitive for a client-side setup. If you're interested I can give you a few more details from my research, but I'm not sure it's a feasible project.

**Comment by interactiveRob** [comment]
For those looking for non-image PDF generation with jsPDF. The current jsPDF documentation (for addHTML) points to this example which uses an older html2pdf library and is able to provide selectable text. https://cdn.rawgit.com/MrRio/jsPDF/master/examples/html2pdf/showcase_supported_html.html

However working with this version, I was unable to get proper page breaks or load custom fonts so I opted for @eKoopmans work instead.

**Comment by Redhaus** [comment]
For those of you that need it. http://pdfmake.org/#/ does a good job of creating lightweight client-side text-based PDF's 
Unfortunately, all of the layouts have to be done in JSON objects and there is no real way I have found to convert an HTML page layout using this library. That said, it works well. Ideally, the best solution would be a hybrid between these two libraries.
